### PR TITLE
Add support for listing files

### LINKED ISSUE #35: Tree#list method for listing files.
Blocked by #25 and #26, possibly others.

Interface something like this:
```ruby
tree = client.tree_connect(...)
tree.list # => [ 'foo.txt', 'bar.bat' ]
```

### PR DESCRIPTION
**Pull request by wvu-r7**
I made a thing. Resolves #35.

**To-do:**

- [x] Add docs
- [ ] Add specs
- [x] Add test script

### CODE REVIEW

**Comment by jlee-r7** [comment]
So with these todo's, are you still working on this, or should I take it?

File: lib/smb2/packet/query.rb [hunk 1]
```diff
@@ -1,6 +1,7 @@
 #
 module Smb2::Packet::Query
   autoload :StandardInformation, 'smb2/packet/query/standard_information'
+  autoload :NamesInformation, 'smb2/packet/query/names_information'
```
Comment (jlee-r7) [inline lib/smb2/packet/query.rb]: Alphabetical order, please.

File: lib/smb2/tree.rb [hunk 2]
```diff
@@ -172,4 +219,26 @@ def disposition_from_file_mode(mode)
     end
   end
 
+  def class_array_from_blob(blob, klass)
+    class_array = []
+    offset = 0
+
+    loop do
+      length = blob[offset, 4].unpack('V').first
```
Comment (jlee-r7) [inline lib/smb2/tree.rb]: `unpack` will work on a string longer than the format.  You can do this directly on blob, without the slicing, i.e.:
```ruby
length = blob.unpack('V').first
```

Comment (wvu-r7) [inline lib/smb2/tree.rb]: Ooh, right. Thank you.

**Comment by wvu-r7** [comment]
Fixed our ``unpack``` regression and added your ```offset``` safeguard. Thanks!

File: lib/smb2/packet/query/names_information.rb [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+
+  default_options endian: 'little'
+
+  unsigned :next_entry_offset, 32
+  unsigned :file_index, 32
+  unsigned :file_name_length, 32
+  rest :raw_file_name
+
+  def file_name
+    raw_file_name[0, file_name_length]
```
Comment (jlee-r7) [inline lib/smb2/packet/query/names_information.rb]:
```ruby
raw_file_name[0, file_name_length].force_encoding("utf-16le").encode("utf-8")
```

Comment (wvu-r7) [inline lib/smb2/packet/query/names_information.rb]: I am using ```encode('utf-8', 'utf-16le')```, since we assume the input is UTF-16LE.

**Comment by wvu-r7** [comment]
Rebased against ```upstream/master```.

File: lib/smb2/tree.rb [hunk 4]
```diff
@@ -122,6 +123,66 @@ def inspect
...
+      create_request.filename = directory.encode('utf-16le')
+    else # y u do dis microsoft
+      create_request.filename = "\x00"
+      create_request.filename_length = 0
+    end
+
+    response = send_recv(create_request)
+    create_response = Smb2::Packet::CreateResponse.new(response)
+
+    unless create_response.header.nt_status.zero?
```
Comment (trosen-r7) [inline lib/smb2/tree.rb]: No more `header` -- just use `create_request.nt_status`

File: lib/smb2/tree.rb [hunk 5]
```diff
@@ -122,6 +123,66 @@ def inspect
...
+      file_name: pattern.encode('utf-16le')
+    )
+
+    class_array = []
+
+    loop do
+      response = send_recv(directory_request)
+      directory_response = Smb2::Packet::QueryDirectoryResponse.new(response)
+
+      unless directory_response.header.nt_status.zero?
```
Comment (trosen-r7) [inline lib/smb2/tree.rb]: No more `header` -- just use `directory_response.nt_status`

File: examples/list_tree.rb [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+  exit 1
+end
+
+d = Smb2::Dispatcher::Socket.connect(host, 445)
+
+c = Smb2::Client.new(dispatcher: d, username: "msfadmin", password: "msfadmin")
+
+c.negotiate
+c.authenticate
+tree = c.tree_connect("\\\\#{host}\\Users")
```
Comment (trosen-r7) [inline examples/list_tree.rb]: Make this hit `C$`

File: examples/list_tree.rb [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+
+tree.list(directory: 'public').each { |f| puts f.file_name }
```
Comment (trosen-r7) [inline examples/list_tree.rb]: Glob something that will always be there, like `"*.exe"`

File: lib/smb2/packet/query/names_information.rb [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+
+  default_options endian: 'little'
+
+  unsigned :next_entry_offset, 32
+  unsigned :file_index, 32
+  unsigned :file_name_length, 32
+  rest :raw_file_name
+
+  def file_name
+    raw_file_name[0, file_name_length].encode('utf-8', 'utf-16le')
```
Comment (trosen-r7) [inline lib/smb2/packet/query/names_information.rb]: Grab this encoding and put it in a constant and then use that constant somewhere. Death to magic strings!

File: lib/smb2/tree.rb [hunk 9]
```diff
@@ -122,6 +123,66 @@ def inspect
...
+    create_request = Smb2::Packet::CreateRequest.new(
+      impersonation: Smb2::Packet::IMPERSONATION_LEVELS[:IMPERSONATION],
+      desired_access: Smb2::Packet::DIRECTORY_ACCESS_MASK[:FILE_LIST_DIRECTORY],
+      share_access: Smb2::Packet::SHARE_ACCESS[:FILE_SHARE_READ],
+      disposition: Smb2::Packet::CREATE_DISPOSITIONS[:FILE_OPEN],
+      create_options: Smb2::Packet::CREATE_OPTIONS[:FILE_DIRECTORY_FILE]
+    )
+
+    if directory
+      create_request.filename = directory.encode('utf-16le')
```
Comment (trosen-r7) [inline lib/smb2/tree.rb]: use constant instead of encoding magic string

File: lib/smb2/tree.rb [hunk 10]
```diff
@@ -122,6 +123,66 @@ def inspect
...
+    create_response = Smb2::Packet::CreateResponse.new(response)
+
+    unless create_response.header.nt_status.zero?
+      raise create_response.inspect
+    end
+
+    directory_request = Smb2::Packet::QueryDirectoryRequest.new(
+      file_info_class: Smb2::Packet::FILE_INFORMATION_CLASSES[type],
+      file_id: create_response.file_id,
+      file_name: pattern.encode('utf-16le')
```
Comment (trosen-r7) [inline lib/smb2/tree.rb]: Death to magic strings

File: lib/smb2/tree.rb (hunk 4 above)
Comment (wvu-r7) [inline lib/smb2/tree.rb]: Done.

File: lib/smb2/tree.rb (hunk 5 above)
Comment (wvu-r7) [inline lib/smb2/tree.rb]: Done.

File: examples/list_tree.rb (hunk 6 above)
Comment (wvu-r7) [inline examples/list_tree.rb]: Done. Also changed in ```examples/read_file.rb```.

File: examples/list_tree.rb (hunk 7 above)
Comment (wvu-r7) [inline examples/list_tree.rb]: Default is ```*```.

**Pull request merged**